src/group_theory/is_free_group.lean

```diff
@@ -18,47 +34,74 @@ universes u w
+    have := free_group.lift.symm.bijective.exists_unique f,
+    simp_rw function.funext_iff at this,
+    exact this,
   end }
 
 namespace is_free_group
 
 variables {G H : Type u} [group G] [group H] [is_free_group G]
 
-/-- The group homomorphism from a free group given by some function on the generators. -/
-def lift' (f : generators G → H) : G →* H := classical.some (unique_lift f)
+/-- The equivalence between functions on the generators and group homomorphisms from a free group
+given by those generators. -/
+@[simps symm_apply]
+def lift' : (generators G → H) ≃ (G →* H) :=
+{ to_fun := λ f, classical.some (unique_lift' f),
+  inv_fun := λ F, F ∘ of,
+  left_inv := λ f, funext (classical.some_spec (unique_lift' f)).left,
+  right_inv := λ F, ((classical.some_spec (unique_lift' (F ∘ of))).right F (λ _, rfl)).symm }
+
 @[simp] lemma lift'_of (f : generators G → H) (a : generators G) : (lift' f) (of a) = f a :=
-(classical.some_spec (unique_lift f)).left a
+congr_fun (lift'.symm_apply_apply f) a
+
+@[simp] lemma lift'_eq_free_group_lift {A : Type u} :
+  (@lift' (free_group A) G _ _ _) = free_group.lift :=
+begin
+  -- TODO: `apply equiv.symm_bijective.injective`,
+  rw [←free_group.lift.symm_symm, ←(@lift' (free_group A) G _ _ _).symm_symm],
+  congr' 1,
+  ext,
+  refl,
+end
+
+@[simp] lemma of_eq_free_group_of {A : Type u} : (@of (free_group A) _ _) = free_group.of :=
+rfl
+
 lemma eq_lift' (f : generators G → H) (F : G →* H) (h : ∀ a, F (of a) = f a) : F = lift' f :=
-(classical.some_spec (unique_lift f)).right F h
+lift'.symm_apply_eq.mp (funext h)
+
+@[ext]
+lemma ext_hom ⦃f g : G →* G⦄ (h : ∀ a, f (of a) = g (of a)) :
+  f = g :=
+lift'.symm.injective $ funext h
 
 lemma end_is_id (f : G →* G) (h : ∀ a, f (of a) = of a) : ∀ g, f g = g :=
-let u := eq_lift' (f ∘ of) in
-have claim : f = monoid_hom.id G := trans (u _ (λ _, rfl)) (u _ (by simp [h])).symm,
-monoid_hom.ext_iff.mp claim
+suffices f = monoid_hom.id G, from monoid_hom.congr_fun this,
+ext_hom h
 
 variable (G)
 /-- Any free group is isomorphic to "the" free group. -/
-@[simps] def iso_free_group_of_is_free_group : G ≃* free_group (generators G) :=
+@[simps] def to_free_group : G ≃* free_group (generators G) :=
 { to_fun := lift' free_group.of,
   inv_fun := free_group.lift of,
-  left_inv := end_is_id ((free_group.lift of).comp (lift' free_group.of)) (by simp),
-  right_inv := begin
-    have : (lift' free_group.of).comp (free_group.lift of) =
-      monoid_hom.id (free_group (generators G)),
-    { ext, simp },
-    rwa monoid_hom.ext_iff at this,
-  end,
+  left_inv :=
+    suffices (free_group.lift of).comp (lift' free_group.of) = monoid_hom.id G,
+    from monoid_hom.congr_fun this,
+    by { ext, simp },
+  right_inv :=
+    suffices
+      (lift' free_group.of).comp (free_group.lift of) = monoid_hom.id (free_group (generators G)),
+    from monoid_hom.congr_fun this,
+    by { ext, simp },
   map_mul' := (lift' free_group.of).map_mul }
 
```

Comment:
It could be nice to know that `of` is always injective. This should follow from `free_group.of_injective`, transporting across `to_free_group`.

Reply:
I think I'll leave that as a follow-up if that's alright.

Reply:
Of course!